src/cpufreq.c

```diff
@@ -25,8 +25,71 @@
 #include "common.h"
 #include "plugin.h"
 
+#define MAX_AVAIL_FREQS 20
+
 static int num_cpu;
 
+struct thread_data {
+  long long time_prev[MAX_AVAIL_FREQS];
+  long long transitions;
+} * t_data;
+
+/* Flags denoting capability of reporting stats. */
+unsigned report_time_in_state, report_total_trans;
+
+static int counter_init(void) {
+  t_data = calloc(num_cpu, sizeof(struct thread_data));
+  if (t_data == NULL)
+    return 0;
+
+  report_time_in_state = 1;
+  report_total_trans = 1;
+
+  /* Initialize time in state counters */
+  for (int i = 0; i < num_cpu; i++) {
+    char filename[256];
+    FILE *fh;
+    int j = 0;
+    char state[DATA_MAX_NAME_LEN], buffer[DATA_MAX_NAME_LEN];
+    long long t;
+
+    snprintf(filename, sizeof(filename),
+             "/sys/devices/system/cpu/cpu%d/cpufreq/stats/time_in_state", i);
+    fh = fopen(filename, "r");
+    if (fh == NULL) {
+      report_time_in_state = 0;
+      return 0;
+    }
+    while (fgets(buffer, sizeof(buffer), fh) != NULL) {
+      if (!sscanf(buffer, "%s%lli", state, &t)) {
+        fclose(fh);
+        return 0;
+      }
+      t_data[i].time_prev[j] = t;
+      j++;
```

Comment:
This cycle does not check for MAX_AVAIL_FREQS limit, so value can exit over array bounds.

Reply:
fixed in next revision of patch